src/nemos/base_regressor.py

```diff
@@ -0,0 +1,418 @@
+
+        Parameters
+        ----------
+        solver_name :
+            Name of the solver.
+        solver_kwargs :
+            Additional keyword arguments for the solver.
+
+        Raises
+        ------
+        NameError
+            If any of the solver keyword arguments are not valid.
+        """
+        solver_args = inspect.getfullargspec(getattr(jaxopt, solver_name)).args
+        undefined_kwargs = set(solver_kwargs.keys()).difference(solver_args)
+        if undefined_kwargs:
+            raise NameError(
+                f"kwargs {undefined_kwargs} in solver_kwargs not a kwarg for jaxopt.{solver_name}!"
+            )
+
+    def instantiate_solver(
+        self, *args, **kwargs
+    ) -> Tuple[SolverInit, SolverUpdate, SolverRun]:
+        """
+        Instantiate the solver with the provided loss function.
+
+        Instantiate the solver with the provided loss function, and return callable functions
+        that initialize the solver state, update the model parameters, and run the optimization.
+
+        This method creates a solver instance from jaxopt library, tailored to the specific loss
+        function and regularization approach defined by the Regularizer instance. It also handles
+        the proximal operator if required for the optimization method. The returned functions are
+        directly usable in optimization loops, simplifying the syntax by pre-setting
+        common arguments like regularization strength and other hyperparameters.
+
+        Parameters
+        ----------
+        loss :
+            The loss function to be optimized.
+
+        *args:
+            Positional arguments for the jaxopt `solver.run` method, e.g. the regularizing
+            strength for proximal gradient methods.
+
+        prox:
+            Optional, the proximal projection operator.
+
+        *kwargs:
+            Keyword arguments for the jaxopt `solver.run` method.
+
+        Returns
+        -------
+        :
+            A tuple containing three callable functions:
+            - solver_init_state: Function to initialize the solver's state, necessary before starting the optimization.
+            - solver_update: Function to perform a single update step in the optimization process,
+            returning new parameters and state.
+            - solver_run: Function to execute the optimization process, applying multiple updates until a
+            stopping criterion is met.
+        """
+        # final check that solver is valid for chosen regularizer
+        if self.solver_name not in self.regularizer.allowed_solvers:
+            raise ValueError(
+                f"The solver: {self.solver_name} is not allowed for "
+                f"{self._regularizer.__class__} regularizaration. Allowed solvers are "
+                f"{self._regularizer.allowed_solvers}."
+            )
+
+        # only use penalized loss if not using proximal gradient descent
+        if self.solver_name != "ProximalGradient":
```

Comment:
if I'm understanding correctly, all regularizers have a `penalized_loss` method, but Lasso and GroupLasso can only use the ProximalGradient solver, so they never get used? is that right?

Reply:
Yes, for now. The goal was to standardize the API to make it agnostic. I think for Lasso and GroupLasso the hope would be that there would be other solvers available in the future that could use the `penalized_loss`. For now, we are using `scipy.optimize.minimize` with the Nelder Mead method to check that the loss functions we have implemented are reaching the same optimum.

Reply:
I think this makes sense, with some additional explanatory comments that Edoardo is adding.